Add a notification and restore function for unexpected exception/error (#1031)
smartdevicelink/sdl_evolution: proposals/NNNN-Add-a-notification-and-restore-function-for-unexpected-exception-error.md

## Proposed solution
We propose the following items to improve the problems mentioned above.
1. Addition of notification process when unexpected exception/error occurs
2. Addition of restore process after unexpected exception/error occcured

#### 1. Addition of notification process when unexpected exception/error occurs
App crash and app kill are the two types of causes for an unexpected exception/error. App crash refers to the unexpected exception/error due to design while launching the app. On the other hand, app kill refers to the swipe out operation by the user or termination by the system while the app is running.

<b>a) App Kill</b>

> - In the case of Android, app kill notifications are monitored with the router service, while it is with the SDL proxy in the case of iOS.

Reviewer: ```suggestion
- In the case of Android, app kill notifications are monitored with the router service, while it is with the SDL app library in the case of iOS.
```

> - SDL proxy sends `UnregisterAppInterface` that includes the cause of the unexpected exception/error to SDL Core.

Reviewer: ```suggestion
- SDL app library sends `UnregisterAppInterface` that includes the cause of the unexpected exception/error to SDL Core.
```

- SDL Core notifies onscreen (ONS) message display to HMI after receiving the app kill notification.
- Then, the HMI displays a pop-up to notify the user with the cause of the unexpected exception/error of the app.
- In the case of app kill, the process of restoring the state before the app was terminated will NOT be perfomed.

<b>b) App Crash</b>

- In case the Android, app crash notifications are monitored with the router service, while it is with the app in the case of iOS.

> - After receiving the app crash notification, the app calls the error handling provided by the SDL proxy and sends the cause of the unexpected exception/error of the app to the SDL proxy.

Reviewer: ```suggestion
- After receiving the app crash notification, the app calls the error handling provided by the SDL app library and sends the cause of the unexpected exception/error of the app to the SDL app library.
```

> - At the time of the app crash, SDL proxy saves the template (`SetDisplayLayout`) displayed on the head unit (HU).

Reviewer: ```suggestion
- At the time of the app crash, SDL app library saves the template (`SetDisplayLayout`) displayed on the head unit (HU).
```

> - SDL proxy sends `UnregisterAppInterface` that includes the cause of the unexpected exception/error to SDL Core.

Reviewer: ```suggestion
- SDL app library sends `UnregisterAppInterface` that includes the cause of the unexpected exception/error to SDL Core.
```

- SDL Core notifies ONS message display to HMI after receiving the app crash notification.

> - Then, the HMI display a pop-up to notify the user the cause of unexpected exception/error of the app.

Reviewer: ```suggestion
- Then, the HMI displays a pop-up to notify the user with the cause of the unexpected exception/error of the app.
```

> - SDL proxy records the flag in preparation for restoring the app.

Reviewer: ```suggestion
- SDL app library records the flag in preparation for restoring the app.
```

> <b>Figure 1.</b> Notification of Unexpected exception/error

Reviewer: ```suggestion
<b>Figure 1.</b> Notification of Unexpected Exception/Error
```

![Figure_1_error_notification_sequence.png](../assets/proposals/NNNN-Add-a-notification-and-restore-function-for-unexpected-exception-error/Figure_1_error_notification_sequence.png)

#### 2. Addition of restore process after unexpected exception/error occcured

> - When SDL proxy (`SdlManager`, `LifeCycleConfigure`) is initialized, it is necessary to set whether or not to restore the app during unexpected exception/error.

Reviewer: ```suggestion
- When SDL proxy (`SdlManager`, `LifeCycleConfigure`) is initialized, it is necessary to set whether or not to restore the app during an unexpected exception/error.
```

> - When SDL proxy (`SdlManager`, `LifeCycleConfigure`) is initialized, it is necessary to set whether or not to restore the app during an unexpected exception/error.

Reviewer: ```suggestion
- When SDL app library (`SdlManager`, `LifeCycleConfigure`) is initialized, it is necessary to set whether or not to restore the app during an unexpected exception/error.
```

> - When the app restarts, SDL proxy checks whether the template displayed on the HU at the time of the app crash is saved.

Reviewer: ```suggestion
- When the app restarts, SDL app library checks whether the template displayed on the HU at the time of the app crash is saved.
```

> - If the template during the app crash is saved, SDL proxy send `Alert` to SDL Core.

Reviewer: ```suggestion
- If the template during the app crash is saved, SDL proxy sends `Alert` to SDL Core.
```

> - If the template during the app crash is saved, SDL proxy sends `Alert` to SDL Core.

Reviewer: ```suggestion
- If the template during the app crash is saved, SDL app library sends `Alert` to SDL Core.
```

- After receiving `Alert`, the HMI displays a pop-up and informs the user of the option to restore.

> - If restore is selected, SDL proxy sends the RPC of the template at the time of app crash, and then the HMI displays that template. If restore is NOT selected, SDL proxy discards the restoration information, and then the app launches as it is.

Reviewer: ```suggestion
- If restore is selected, SDL proxy sends the RPC of the template at the time of app crash, and then the HMI displays that template. If restore is NOT selected, SDL proxy discards the restoration information, and then the app launches as is.
```

> - If restore is selected, SDL proxy sends the RPC of the template at the time of app crash, and then the HMI displays that template. If restore is NOT selected, SDL proxy discards the restoration information, and then the app launches as is.

Reviewer: ```suggestion
- If restore is selected, SDL app library sends the RPC of the template at the time of app crash, and then the HMI displays that template. If restore is NOT selected, SDL app library discards the restoration information, and then the app launches as is.
```

> - During the restoration, the HMI display a "Restoration in Progress" pop-up, while SDL proxy performs the restore process by rejecting all RPCs sent by the app.

Reviewer: ```suggestion
- During the restoration, the HMI displays a "Restoration in Progress" pop-up, while SDL proxy performs the restore process by rejecting all RPCs sent by the app.
```

> - During the restoration, the HMI displays a "Restoration in Progress" pop-up, while SDL proxy performs the restore process by rejecting all RPCs sent by the app.

Reviewer: ```suggestion
- During the restoration, the HMI displays a "Restoration in Progress" pop-up, while SDL app library performs the restore process by rejecting all RPCs sent by the app.
```

> - After the restoration is complete, SDL Proxy sends the restored template to App.

Reviewer: ```suggestion
- After the restoration is complete, SDL Proxy sends the restored template to the app.
```

> - After the restoration is complete, SDL Proxy sends the restored template to the app.

Reviewer: ```suggestion
- After the restoration is complete, SDL app library sends the restored template to the app.
```

- Finally, the app proceeds to the next operation after receiving the restored template.

<b>Figure 2.</b> Restore Process of the App

![Figure_2_return_processing.png](../assets/proposals/NNNN-Add-a-notification-and-restore-function-for-unexpected-exception-error/Figure_2_return_processing.png)

The contents suggested by this proposal are as follows:

i) Add an attribute, `unregisterReason`,  to `UnregisterAppInterface` on Mobile_API and `OnAppUnregistered` on HMI_API.

<b>Mobile_API.xml</b>

```xml
    <function name="UnregisterAppInterface" functionID="UnregisterAppInterfaceID" messagetype="request" since="1.0">
        <description>
            Closes an interface from a mobile application.
            After unregisterAppInterface, no commands other than registerAppInterface will be accepted/executed.
            Will fail, if no registerAppInterface was completed successfully before.
        </description>
 +      <param name="unregisterReason" type="ReasonType" platform="documentation" mandatory="true">

>  +          <description>Specifies reason for unregistering the apllication.</description>

Reviewer: ```suggestion
 +          <description>Specifies reason for unregistering the application.</description>
```

 +      </param>
    </function>

 +  <enum name="ReasonType">
 +      <description>Describes the reasons for unregistering the application.</description>
 +      <element name="APP_CRASH">
 +          <description>Indicates the unexpected exception/error due to design while launching the application.</description>
 +      </element>
 +      <element name="APP_KILL">
 +          <description>Indicates the swipe out operation by user or termination by the system while the application is running.</description>
 +      </element>
 +  </enum>
```

<b>HMI_API.xml</b>

```xml
    <function name="OnAppUnregistered" messagetype="notification">
        <description>
            Issued by SDL to notify HMI about application unregistered.
            Application then to be removed from application list; all data connected with application has to be cleared up.
        </description>
 +      <param name="unregisterReason" type="Common.ReasonType" mandatory="true">

>  +          <description>Specifies reason for unregistering the apllication.</description>

Reviewer: ```suggestion
 +          <description>Specifies reason for unregistering the application.</description>
```

 +      </param>
    </function>

 +  <enum name="ReasonType">
 +      <description>Describes the reasons for unregistering the application.</description>
 +      <element name="APP_CRASH">
 +          <description>Indicates the unexpected exception/error due to design while launching the application.</description>
 +      </element>
 +      <element name="APP_KILL">
 +          <description>Indicates the swipe out operation by user or termination by the system while the application is running.</description>
 +      </element>
 +  </enum>
```

ii) Add an attribute, `isReset`, to `SDLLifecycleConfiguration`, which determines whether to restore the app or not.<br>
iii) Add a backup process that saves the template at the time of App Crash to `SdlManager` and send `UnregisterAppInterface`.<br>

> iv) SDL proxy checks whether the restore information is saved when `isReset` is set to true. If there is information, SDL proxy sends `Alert` to the HMI. When the user selects to restore on the HMI, SDL proxy sends the template for app crash to the HMI.

Reviewer: ```suggestion
iv) SDL app library checks whether the restore information is saved when `isReset` is set to true. If there is information, SDL app library sends `Alert` to the HMI. When the user selects to restore on the HMI, SDL app library sends the template for app crash to the HMI.
```

## Potential downsides
None.

## Impact on existing code

> - The addition of a new attribute to `UnregisterAppInterface` affects SDL Core and SDL proxy.

Reviewer: ```suggestion
- The addition of a new attribute to `UnregisterAppInterface` affects SDL Core and SDL app library.
```

- The addition of a new attribute to `OnAppUnregistered` affects SDL Core and HMI.

> - SDL proxy must monitor app kill notifications and provide processing during unexpected exception/error.

Reviewer: ```suggestion
- SDL app library must monitor app kill notifications and provide processing during unexpected exception/error.
```
